File: CobwebApp/src/main/java/cobweb3d/core/agent/Agent.java
package cobweb3d.core.agent;


import cobweb3d.core.Updatable;
import cobweb3d.core.location.LocationDirection;

/**
 * Basic properties of an Agent
 */
public abstract class Agent implements Updatable {
    public LocationDirection position;
    private Integer id;
    private int type;
    private boolean alive = true;
    private int energy;

    public Agent(int type) {
        this.type = type;
    }

    public int id() {
        return id;
    }

    public void setId(Integer id) {
        if (this.id != null) throw new IllegalStateException("Agent id is already set for " + this.toString() + "!");
        else this.id = id;
    }

    public void die() {
        assert (isAlive());
        if (!isAlive())
            return;
        alive = false;
    }

    public int getEnergy() {
        return energy;
    }

    public boolean enoughEnergy(int required) {
        return getEnergy() >= required;
    }

    /**
     * Changes the agent's energy level.
     *
     * @param delta Energy change delta, positive means agent gains energy, negative means it loses
     * @param cause Why the energy changed.
     */
    public void changeEnergy(int delta) {
        energy += delta;
    }

    /**
     * @return the location this Agent occupies.
     */
    public LocationDirection getPosition() {
        return position;
    }

    public boolean isAlive() {
        return alive;
    }

    public int getType() {
        return type;
    }

    protected abstract Agent createChildAsexual(LocationDirection location);
}

File: src/main/java/cobweb3d/model/agent/Agent.kt
package cobweb3d.model.agent

import cobweb3d.model.Updatable
import cobweb3d.model.entity.Cause
import cobweb3d.model.location.LocationDirection

/**
 * Basic properties of an agent
 */
abstract class Agent(val type: Int) : Updatable {

    var id: Int = 0

    var isAlive = true // Initializer required, not a nullable type.
        private set // The setter is private and has the default implementation.

    /**
     * @return the location this agent occupies.
     */
    var position: LocationDirection? = null
        protected set

    /**
     * Energy the agent can use to do things and can gain doing other things
     */
    var energy: Int = 0
        private set

    fun die() {
        assert(isAlive)
        if (!isAlive)
            return

        isAlive = false
    }

    fun enoughEnergy(required: Int): Boolean {
        return energy >= required
    }

    /**
     * Changes the agent's energy level.
     * @param delta Energy change delta, positive means agent gains energy, negative means it loses
     * @param cause Why the energy changed.
     */
    fun changeEnergy(delta: Int, cause: Cause) {
        energy += delta
    }

    protected abstract fun createChildAsexual(location: LocationDirection): Agent

    override fun update() {}
}
